implement plugin system for adding complex functionality

bezmi opened the pull request:
The additions here allow loading of plugins. See `src/CircleTestPlugin.js`  for one such plugin. Having it as a factory function allows a lot of flexibility. I think this implementation is reasonably 'safe' as plugins are given only the map object and the options from the config. Below are some instructions on how to get it working for testing. Documentation and commenting in the classes I've added needs work. I tried to follow your structure @nathan-gs.

# Installation
1. Build and install the map card
2. Copy `src/CircleTestPlugin.js` to the `www` folder.

# Testing
Here is a config for lovelace that will draw two circles on the map.
```yaml
type: custom:map-card
x: -23.698
"y": 133.8807
zoom: 1
debug: true
plugins:
  - name: circle1
    url: /local/CircleTestPlugin.js
    options:
      x: -23.698
      "y": 133.8807
      r: 3000000
      fillColor: green
  - name: circle2
    url: /local/CircleTestPlugin.js
    options:
      x: 20
      "y": 100
      r: 1000000
      fill: false
      color: orange
```
## Result
![circle_plugin](https://github.com/user-attachments/assets/dbdf9d2d-e50e-49b8-961f-3d9620c39f0a)

# Other thoughts
Some of the methods I added are async. I did notice that there are other async methods such as those in `TileLayersRenderService.js` as well. Are these being used correctly if the setup functions in `MapCard.js` aren't themselves async? 

# Backstory
The Aussie Bureau of Meteorology have an undocumented API that serves mapbox vector tiles in their mobile app. I like the BOM radar better than the rainviewer one that I have been using, as it is much higher resolution. I spent some time working on an implementation that can read the vector tiles using `MapLibre GL JS` and plot it on a leaflet map via `maplibre-gl-leaflet`.  Unfortunately, the way they have set up their vector tiles is not exactly 'standard'. It's one of those things that's extremely inefficient and complicated to do if you use homeassistant template sensors and the built-in url templating functionality only, but can be done with only a few lines of code. I think that it makes a decent case for 'plugins' to exist.

nathan-gs commented:
hey @bezmi thanks for the PR! I like the idea, although I hope most things will end-up inside this repo (but from the backstory there's a good case for plugins as well). 

I spend yesterday on the Circle feature, including some refactors and a lot of extra tests, will launch the PR for that one soon.

nathan-gs commented on src/services/render/PluginsRenderService.js:
```diff
@@ -0,0 +1,83 @@
+    }
+  }
+
+  async render() {}
```
I believe we need to call the ` render` methods of the plugins here.

bezmi replied:
I call `render` in setup, as it is the same as `TileLayersRenderService` calling `layer.render()`. If I loop over `pluginInstance.render()`, it is called repeatedly via `MapCard.render()` since it relates to the html template in Lit elements. Perhaps there should be more distinction between `render` as it relates to the template of the web component and `render` as it relates to drawing on the map?

bezmi replied:
I don't think the plugins would be modifying html directly in any way, other than drawing on the map, so maybe remove the `PluginsRenderService.render()` call inside `MapCard.render()` and then rename `Plugin.render()` to `Plugin.renderMap()`?

nathan-gs commented on src/CircleTestPlugin.js:
Let's have a dedicated `plugins/` folder under the root of the repo.

bezmi replied:
good idea

nathan-gs commented on src/components/MapCard.js:
```diff
@@ -43,33 +44,39 @@ export default class MapCard extends LitElement {
   entitiesRenderService;
   /** @type {InitialViewRenderService} */
   initialViewRenderService;
+  /** @type {PluginManager} */
```
should be `PluginsRenderService`

bezmi commented:
@nathan-gs I have pushed some changes that deal with my comments. Let me know what you think, happy to revert anything if need be.

bezmi commented:
Yes, I'm still testing out the interface with the more complex weather radar plugin to see if it needs any changes before merging. Once I'm happy with it, I'll add some tests.

Kevinjil commented:
This is exactly what I've been looking for for some time. I've created a plugin for my personal instance, and to try this interface: https://github.com/Kevinjil/ha-map-card-buienradar

bezmi commented:
I have made some changes to the interface. Working plugin for the australian BoM data here: https://github.com/bezmi/ha-map-card-plugin-bom-radar. The sample config should work.

- plugin factory function now requires `Logger` as argument. Keeps things neat.
- added `cleanup()` to `PluginsRenderService` and `destroy()` to `Plugin`. I needed these to clear an interval/timeout that was running an update loop and causing an error when the page was going into 'edit' mode and back, as they kept running even though the rest of the card was reloaded. I think they are important additions.

Next step: docs and tests (fun!)

bezmi commented:
@nathan-gs I have added tests, adjusted some of the logic, added a type definition for typescript plugins and updated the README with details on the plugin system. Functionality is all there, ready for merge pending your review.

@Kevinjil I have put a link to your plugin in the README. Let me know if you would like this removed.

nathan-gs commented:
Thank you @bezmi! I will review and test it out.

nathan-gs commented on src/components/Plugin.js:
is this deliberately empty?

Kevinjil commented:
> @Kevinjil I have put a link to your plugin in the README. Let me know if you would like this removed.

I'm planning on keeping the plugin in a working state (and make it installable with HACS once this is merged), so people are free to grab and use it.

nathan-gs commented:
@Kevinjil very cool! Will see if works with buienradar.be as well!

Kevinjil commented:
> @Kevinjil very cool! Will see if works with buienradar.be as well!

I can probably add some configuration to select `nl`, `be`, `eu` radar images.

nathan-gs commented on src/models/Plugin.js:
```diff
@@ -0,0 +1,31 @@
+    // useful if plugin needs to respond to HA state.
+  }
+
+  destroy() {
```
should we make this mandatory?

nathan-gs commented on README.md:
```diff
@@ -179,6 +180,38 @@ More complex use of the WMS/Tile history property can be configured within the h
+| [`bom-radar`](https://github.com/bezmi/ha-map-card-plugin-bom-radar) | Displays the Australian BoM rainfall radar for the past 90 minutes and the radar forecast for the next 90 minutes as an overlay on the map. | 
+| [`buienradar`](https://github.com/Kevinjil/ha-map-card-buienradar) | Displays `buienradar.nl` as an overlay on the map | 
+
+
```
Let's add a sentence:
```
You can find more plugins using the [ha-map-card-plugin](https://github.com/topics/ha-map-card-plugin) topic.
```

nathan-gs commented on README.md:
```diff
@@ -198,6 +231,10 @@ This project uses [devenv.sh](https://devenv.sh/).
+* All plugins should implement the `Plugin` class. See [`Plugin.js`](./src/models/Plugin.js) and [`Plugin.d.ts`](./src/models/Plugin.d.ts).
+* For a concrete example, see [`CircleTestPlugin.js`](./plugins/CircleTestPlugin.js).
+* Typescript example: see [bezmi/ha-map-card-plugin-bom-radar](https://github.com/bezmi/ha-map-card-plugin-bom-radar).
 
```
Add a sentence:
```
Tag your Github repo with [ha-map-card-plugin](https://github.com/topics/ha-map-card-plugin) for discoverability.
```

nathan-gs commented:
In release [v1.13.0](https://github.com/nathan-gs/ha-map-card/releases/tag/v1.13.0), thanks again!
